web/interceptor/access_control.go

```diff
@@ -0,0 +1,176 @@
+package interceptor
+
+import (
+	"context"
+	"strings"
+
+	"github.com/quickfeed/quickfeed/qf"
+	"github.com/quickfeed/quickfeed/web/auth"
+	"go.uber.org/zap"
+	"google.golang.org/grpc"
+)
+
+type (
+	role      int
+	roles     []role
+	requestID interface {
+		IDFor(string) uint64
+	}
+)
+
+const (
```

Comment:
Not sure if my rephrasing is accurate, please check. I prefer to line them up with comments on the rhs. You should read it left to right, as in "_none can be anyone_", "_user can access ..._ ", "_group user is enrolled_ ..." and so on.
```go
const (
	none        role = iota // can be anyone.
	user                    // can access its own information.
	group                   // user is enrolled in the course and is member of the given group.
	student                 // user is enrolled in the course with any role.
	teacher                 // user is enrolled in the course with teacher status.
	courseAdmin             // user is enrolled in the course and is also admin.
	admin                   // user has admin privileges.
)
```
Does it make sense?

I'm wondering, do we need the `courseAdmin` role? Is it the same as the `CourseCreatorID`? How do you get this role?

Reply:
It is this role from the original access control:
```
// Access policy:
// Admin enrolled in CourseID,
```

Reply:
Yeah, I just don't remember why it is needed.
```go
// ensure that current user is teacher or enrolled admin to process the submission request
if !s.hasCourseAccess(usr.GetID(), in.GetCourseID(), func(e *qf.Enrollment) bool {
	switch e.Status {
	case qf.Enrollment_TEACHER:
		return true
	case qf.Enrollment_STUDENT:
		return usr.IsAdmin
	}
	return false
}) {
```
So basically, it is a user enrolled as a student in the course. That student also has `admin` privileges. When is this scenario relevant?

The `courseAdmin` users can only use:
- GetSubmissions
- GetSubmissionsByCourse

I'm wondering if this is some legacy issue that is no longer relevant.

@JosteinLindhom Is this something that the frontend needs?

Reply:
I don't remember why it existed in the first place, but I think the `courseAdmin` role is not really needed for `GetSubmissions`, at least I can't see what use case it could break. 

But `GetSubmissionsByCourse` is only used by the `approvelist` tool, what kind of permissions are needed there? Should a non-teacher be able to use it?

Reply:
`GetSubmissionsByCourse` is also used by the frontend (actions.tsx):
```typescript
    const result = await effects.grpcMan.getSubmissionsByCourse(courseID, SubmissionsForCourseRequest.Type.ALL, true)
    const groups = await effects.grpcMan.getSubmissionsByCourse(courseID, SubmissionsForCourseRequest.Type.GROUP, true)
```
But `approvelist` should only be needed by a teacher.